File: python/src/atlassian_sdk/jira/schemas.py
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field


def text_to_adf(text: str) -> dict:
    """Wrap plain text in Atlassian Document Format."""
    return {
        "type": "doc",
        "version": 1,
        "content": [{"type": "paragraph", "content": [{"type": "text", "text": text}]}],
    }


class CreateIssueRequest(BaseModel):
    """Input schema for creating a Jira issue."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    summary: str
    description: str | None = None
    issue_type: str = Field(default="Task", alias="type")
    assignee: str | None = None
    labels: list[str] = Field(default_factory=list)
    parent: str | None = None

    def to_request_body(self, project: str) -> dict:
        """Build the REST API v3 request body.

        Args:
            project: The Jira project key.

        Returns:
            A dict suitable for ``POST /rest/api/3/issue``.
        """
        fields: dict = {
            "project": {"key": project},
            "summary": self.summary,
            "issuetype": {"name": self.issue_type},
        }
        if self.description is not None:
            fields["description"] = text_to_adf(self.description)
        if self.assignee is not None:
            fields["assignee"] = {"id": self.assignee}
        if self.labels:
            fields["labels"] = self.labels
        if self.parent is not None:
            fields["parent"] = {"key": self.parent}
        return {"fields": fields}


_DEFAULT_SEARCH_FIELDS = [
    "summary", "status", "issuetype", "priority",
    "assignee", "creator", "reporter", "project",
    "labels", "created", "updated", "parent",
    "duedate", "resolution", "description",
]


class SearchIssuesRequest(BaseModel):
    """Input schema for searching Jira issues via JQL."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    jql: str
    limit: int = 50
    fields: list[str] | None = None

    def to_request_body(self) -> dict:
        """Build the REST API request body.

        Returns:
            A dict suitable for ``POST /rest/api/3/search/jql``.
        """
        return {
            "jql": self.jql,
            "maxResults": self.limit,
            "fields": self.fields if self.fields is not None else _DEFAULT_SEARCH_FIELDS,
        }


class TransitionIssueRequest(BaseModel):
    """Input schema for transitioning a Jira issue."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True)

    transition_id: str

    def to_request_body(self) -> dict:
        """Build the REST API request body.

        Returns:
            A dict suitable for ``POST /rest/api/3/issue/{key}/transitions``.
        """
        return {"transition": {"id": self.transition_id}}

File: node/src/jira/schemas.ts
/**
 * Zod schemas for Jira input validation with toRequestBody() helpers.
 */

import { z } from "zod";

/** Wrap plain text in Atlassian Document Format (ADF). */
export function textToAdf(text: string): Record<string, unknown> {
  return {
    type: "doc",
    version: 1,
    content: [{ type: "paragraph", content: [{ type: "text", text }] }],
  };
}

export const createIssueSchema = z
  .object({
    projectKey: z.string().optional(),
    summary: z.string(),
    issueTypeName: z.string().default("Task"),
    description: z.string().optional(),
    assigneeId: z.string().optional(),
    labels: z.array(z.string()).optional(),
    parentKey: z.string().optional(),
    epic: z.string().optional(),
  })
  .strict();

export type CreateIssueInput = z.infer<typeof createIssueSchema>;

/** Build the REST API request body for creating an issue. */
export function createIssueBody(
  input: CreateIssueInput & { resolvedProjectKey: string; resolvedParentKey?: string },
): Record<string, unknown> {
  const fields: Record<string, unknown> = {
    project: { key: input.resolvedProjectKey },
    summary: input.summary,
    issuetype: { name: input.issueTypeName },
  };
  if (input.description) fields.description = textToAdf(input.description);
  if (input.assigneeId) fields.assignee = { id: input.assigneeId };
  if (input.labels?.length) fields.labels = input.labels;
  if (input.resolvedParentKey) fields.parent = { key: input.resolvedParentKey };
  return { fields };
}

export const searchIssuesSchema = z
  .object({
    jql: z.string(),
    maxResults: z.number().int().positive().default(50),
    fields: z.array(z.string()).optional(),
  })
  .strict();

export type SearchIssuesInput = z.infer<typeof searchIssuesSchema>;

/** Build the REST API request body for searching issues. */
export function searchIssuesBody(
  input: SearchIssuesInput,
): Record<string, unknown> {
  const body: Record<string, unknown> = {
    jql: input.jql,
    maxResults: input.maxResults,
  };
  if (input.fields?.length) body.fields = input.fields;
  return body;
}

export const transitionIssueSchema = z
  .object({
    transitionId: z.string(),
  })
  .strict();

export type TransitionIssueInput = z.infer<typeof transitionIssueSchema>;

/** Build the REST API request body for transitioning an issue. */
export function transitionIssueBody(
  input: TransitionIssueInput,
): Record<string, unknown> {
  return { transition: { id: input.transitionId } };
}
